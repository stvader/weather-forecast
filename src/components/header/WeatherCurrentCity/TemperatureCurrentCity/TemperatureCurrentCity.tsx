import React from 'react';
import { Typography } from 'antd';

const { Title } = Typography;

const TemperatureCurrentCity: React.FC<any> = ({ temperature }: any) => (
  <Title level={3}>{temperature}</Title>
);

export default TemperatureCurrentCity;
