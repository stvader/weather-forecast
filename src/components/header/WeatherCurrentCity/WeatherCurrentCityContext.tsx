import { createContext } from 'react';

export const WeatherCurrentContext = createContext<any>(null);

export const WeatherCurrentContextProvider = WeatherCurrentContext.Provider;
