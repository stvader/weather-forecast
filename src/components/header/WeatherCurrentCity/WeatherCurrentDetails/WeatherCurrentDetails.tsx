import React from 'react';

import { Table } from 'antd';
import { ColumnProps } from 'antd/es/table';

import { ITableColumn, ITableData } from '.';

interface IWeatherCurrentDetailsView {
  columns: ColumnProps<ITableColumn>[];
  data: ITableData;
}

const WeatherCurrentDetails: React.FC<IWeatherCurrentDetailsView> = ({
  data,
  columns,
}: IWeatherCurrentDetailsView) => (
  <Table columns={columns} dataSource={data} showHeader pagination={false} bordered size="small" />
);

export default WeatherCurrentDetails;
