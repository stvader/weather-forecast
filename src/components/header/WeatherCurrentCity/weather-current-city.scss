.current-weather-block {
  width: 800px;
  display: flex;
  justify-content: space-around;
}
