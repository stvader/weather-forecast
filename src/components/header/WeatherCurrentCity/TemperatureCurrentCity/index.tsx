import React from 'react';
import { connect } from 'react-redux';

import { IAppState } from '../../../../reducers/types';
import { selectTempCurrent } from '../../../../selectors';
import TemperatureCurrentCity from './TemperatureCurrentCity';
import { getTemperatureView, ITemperatureView } from '../../../../utils/getTemperatureView';

interface ITemperatureCurrentCity {
  temperatureData: ITemperatureView;
}

const TemperatureCurrentCityContainer: React.FC<ITemperatureCurrentCity> = ({
  temperatureData,
}: ITemperatureCurrentCity) => {
  const temperature = getTemperatureView(temperatureData);

  return <TemperatureCurrentCity temperature={temperature} />;
};

const mapStateToProps = (state: IAppState) => ({
  temperatureData: selectTempCurrent(state),
});

export default connect(mapStateToProps)(TemperatureCurrentCityContainer);
