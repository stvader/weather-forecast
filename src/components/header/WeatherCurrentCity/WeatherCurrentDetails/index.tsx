import React, { useContext, ReactElement } from 'react';

import { ColumnProps } from 'antd/es/table';

import { WeatherCurrentContext } from '../WeatherCurrentCityContext';
import WeatherCurrentDetails from './WeatherCurrentDetails';
import WindCityWeather from '../../../WindCityWeather';

export interface ITableColumn {
  key: string;
  wind: ReactElement;
  cloudiness: string;
  pressure: string;
  humidity: string;
  sunrise: string;
  sunset: string;
  coords: string;
}

export type ITableData = ITableColumn[];

const columns: ColumnProps<ITableColumn>[] = [
  {
    title: 'Wind',
    dataIndex: 'wind',
    key: 'wind',
  },
  {
    title: 'Cloudiness',
    dataIndex: 'cloudiness',
    key: 'cloudiness',
  },
  {
    title: 'Pressure',
    dataIndex: 'pressure',
    key: 'pressure',
  },
  {
    title: 'Humidity',
    dataIndex: 'humidity',
    key: 'humidity',
  },
  {
    title: 'Sunrise',
    dataIndex: 'sunrise',
    key: 'sunrise',
  },
  {
    title: 'Sunset',
    dataIndex: 'sunset',
    key: 'sunset',
  },
  {
    title: 'Geo coords',
    dataIndex: 'coords',
    key: 'coords',
  },
];

const WeatherCurrentDetailsContainer: React.FC<any> = () => {
  const {
    wind: { direction, speed },
    cloudiness,
    pressure,
    humidity,
    sunrise,
    sunset,
    coord: { lon, lat },
  } = useContext(WeatherCurrentContext);

  const data: ITableData = [
    {
      key: '1',
      wind: <WindCityWeather direction={direction} speed={speed} />,
      cloudiness,
      pressure: `${pressure} hpa`,
      humidity: `${humidity}%`,
      sunrise,
      sunset,
      coords: `[${lon}, ${lat}]`,
    },
  ];

  return <WeatherCurrentDetails data={data} columns={columns} />;
};

export default WeatherCurrentDetailsContainer;
