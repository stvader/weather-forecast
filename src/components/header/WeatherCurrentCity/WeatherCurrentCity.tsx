import React, { useContext } from 'react';
import { Typography } from 'antd';

import TemperatureCurrentCity from './TemperatureCurrentCity';
import WeatherCurrentDetails from './WeatherCurrentDetails';
import { WeatherCurrentContext } from './WeatherCurrentCityContext';

import { bc } from '../../../utils/bem-cn';

import './weather-current-city.scss';

const { Title } = Typography;

const b = bc('current-weather-block');

const WeatherCurrentCity: React.FC = () => {
  const { placeName, country } = useContext(WeatherCurrentContext);

  return (
    <div className={b()}>
      <div className={b('main-info-block')}>
        <Title level={4}>{`${placeName}, ${country}`}</Title>
        <TemperatureCurrentCity />
      </div>

      <WeatherCurrentDetails />
    </div>
  );
};

export default WeatherCurrentCity;
