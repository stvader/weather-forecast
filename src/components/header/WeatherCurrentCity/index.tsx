import React, { useEffect, useMemo } from 'react';
import { connect } from 'react-redux';
import { Spin } from 'antd';

import { IAppState, IWeatherCurrentCity } from '../../../reducers/types';
import { IWeatherCurrentPlace } from '../../../thunks/transformers/transformerWeatherCurrentPlace';
import { getCoordsCurrentPlace } from '../../../thunks';
import { selectWeatherCurrent } from '../../../selectors';
import WeatherCurrentCity from './WeatherCurrentCity';
import { WeatherCurrentContextProvider } from './WeatherCurrentCityContext';

interface IWeatherCurrentCityContainer {
  dispatchGetCoordsCurrentPlace: () => void;
  weatherCurrentCity: IWeatherCurrentCity;
}

interface IWeatherViewWrapper {
  data: IWeatherCurrentPlace | undefined;
}

const WeatherViewWrapper: React.FC<IWeatherViewWrapper> = ({ data }: IWeatherViewWrapper) => (
  <WeatherCurrentContextProvider value={data}>
    <WeatherCurrentCity />
  </WeatherCurrentContextProvider>
);

const WeatherCurrentCityContainer: React.FC<IWeatherCurrentCityContainer> = ({
  weatherCurrentCity: { data, loading, error },
  dispatchGetCoordsCurrentPlace,
}: IWeatherCurrentCityContainer) => {
  useEffect(() => {
    dispatchGetCoordsCurrentPlace();
  }, [dispatchGetCoordsCurrentPlace]);

  const weatherViewMemo = useMemo(() => <WeatherViewWrapper data={data} />, [data]);

  if (loading || !data) {
    return (
      <div>
        <Spin />
      </div>
    );
  }

  if (error) {
    return <div>Ooops... Something wrong</div>;
  }

  return <div>{weatherViewMemo}</div>;
};

const mapStateToProps = (state: IAppState) => ({
  weatherCurrentCity: selectWeatherCurrent(state),
});

const mapDispatchToProps = (dispatch: any) => ({
  dispatchGetCoordsCurrentPlace: () => {
    dispatch(getCoordsCurrentPlace());
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(WeatherCurrentCityContainer);
