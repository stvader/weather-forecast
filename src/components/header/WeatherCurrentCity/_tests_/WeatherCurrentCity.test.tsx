import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Typography } from 'antd';
import WeatherCurrentCity from '../WeatherCurrentCity';
import WeatherCurrentDetails from '../WeatherCurrentDetails';
import { WeatherCurrentContextProvider } from '../WeatherCurrentCityContext';
import TemperatureCurrentCity from '../TemperatureCurrentCity';

const { Title } = Typography;

configure({ adapter: new Adapter() });

const data = 'any';

describe('WeatherCurrentCity', () => {
  it('WeatherCurrentCity render', () => {
    const component = shallow(
      <WeatherCurrentContextProvider value={data}>
        <WeatherCurrentCity />
      </WeatherCurrentContextProvider>,
    );

    const weatherBlock = component.find('.current-weather-block');
    const mainBlock = weatherBlock.find('.current-weather-bloc__main-info-block');
    const details = weatherBlock.find(WeatherCurrentDetails);
    const title = mainBlock.find(Title);
    const tempBlock = mainBlock.find(TemperatureCurrentCity);

    expect(component).toMatchSnapshot();
    expect(weatherBlock).toBeDefined();
    expect(mainBlock).toBeDefined();
    expect(details).toBeDefined();
    expect(title).toBeDefined();
    expect(tempBlock).toBeDefined();
  });
});
