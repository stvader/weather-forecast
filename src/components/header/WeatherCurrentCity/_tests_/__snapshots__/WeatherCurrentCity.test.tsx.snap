// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherCurrentCity WeatherCurrentCity render 1`] = `ShallowWrapper {}`;
